### ISSUE TITLE
Missing x86 vendor intrinsics (SSE2, SSE 4.1, AVX2)

### ISSUE THREAD

**Issue by newpavlov** [rust-lang/stdarch#1178]
Previous issue: #40

# AVX2
 * [ ]  [`_mm256_stream_load_si256`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm256_stream_load_si256&expand=5236)
* [ ]  [`_mm_broadcastsi128_si256`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_broadcastsi128_si256&expand=5236)

# MMX 
EDIT(@workingjubilee): Direct MMX support is no longer in scope for `std::arch`, see:
- https://github.com/rust-lang/stdarch/pull/890

# SSE
* [ ]  [`_mm_free`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_free&expand=5236)
* [ ]  [`_mm_storeu_si16`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_storeu_si16&expand=5236)
* [ ]  [`_mm_loadu_si16`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_loadu_si16&expand=5236)
* [ ]  [`_mm_malloc`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_malloc&expand=5236)
* [ ]  [`_mm_storeu_si64`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_storeu_si64&expand=5236)

# SSE2
* [ ]  [`_mm_loadu_si32`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_loadu_si32&expand=5236)
* [ ]  [`_mm_storeu_si32`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_storeu_si32&expand=5236)

# SSE4.1

* [ ]  [`_mm_stream_load_si128`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_stream_load_si128&expand=5236)

Personally I am interested only in `_mm_stream_load_si128` and `_mm256_stream_load_si256`, but I think it's worth to properly track all unimplemented intrinsics. Some of those intrinsics (e.g. `_mm_malloc` and `_mm_free`) probably should not be exposed, but, in my opinion, motivation behind such decision should be explicitly recorded somewhere (ideally in comments of relevant source files).

**Comment by Lokathor** [comment]
I was under the impression that we'd _deliberately_ removed the MMX stuff.

**Comment by bjorn3** [comment]
Indeed. It requires special handling in the compiler to emit the right type for MMX vectors as they are a different type from regular vectors. In addition it is pretty much impossible to use correctly as LLVM can reorder MMX usage before the intrinsic that enables MMX.

**Comment by newpavlov** [comment]
What about the streaming load intrinsics? Is there a reason why they have been omitted?

**Comment by Lokathor** [comment]
Some of the streaming ops are already in, and stabilized (eg: [_mm_stream_pd](https://doc.rust-lang.org/core/arch/x86_64/fn._mm_stream_pd.html)).

Given this, I'd guess that any missing streaming ops are *likely* an oversight (at least for 128 or 256 bit).

**Comment by jhorstmann** [comment]
`_mm_broadcastsi128_si256` seems to be an alias for `_mm256_broadcastsi128_si256` which is implemented. The intrinsics guide lists both as translating to the same instruction and with the same description.

**Comment by workingjubilee** [comment]
`_mm_malloc` and `_mm_free` seem like they require implementing in libstd.

**Comment by Nilstrieb** [comment]
note that there are issues with these streaming intrinsics, as they have nontemporal hints that are not properly modelled.
: https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/Non-temporal.20stores

**Comment by workingjubilee** [comment]
They've been converted into assembly.

**Comment by sayantn** [comment]
All of these has been implemented, except for `_mm_malloc` and `_mm_free`. They can't really be part of `core::arch`, and there is really no utility of having them over just `std::alloc`
